rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,443 @@
+```
+
+A more fine-grained rule, targeting a specific set of tables:
+
+```yaml
+kind: db_object_import_rule
+version: v1
+metadata:
+  name: import_specific_tables
+spec:
+  db_labels:
+    env: dev
+  mappings:
+    - add_labels:
+        custom_label: my_custom_value
+        env: dev
+      match:
+        procedure_names:
+          - ''
+        table_names:
+          - table1
+          - table2
+          - table3
+        view_names:
+          - ''
+      scope:
+        schema_names:
+          - public
+  priority: 30
+```
+
+#### Import process
+
+The database objects are imported by the database agent when establishing a new
+user session to the database. In this context, they are immediate inputs for
+permission calculation; the desired per-object permissions are subsequently
+written back to the database.
+
+Imported objects are also stored in the backend, where TAG can access them. This
+enables the TAG to visualise the permissions.
+
+Additionally, the imports will be done on a predetermined schedule (e.g. every
+10 minutes), and stored in the backend. If the database engine supports it, the
+sync may also happen when a schema change is detected. For example, in Postgres
+we case use
+[trigger+notify](https://medium.com/launchpad-lab/postgres-triggers-with-listen-notify-565b44ccd782).
+
+#### Import result: the `db_object` resource
+
+Import process creates a number of `db_object` resources with applied labels.
+Aside from standard metadata fields, the object spec consists of a number of
+predefined attributes:
+
+```protobuf
+// DatabaseObjectSpec is the spec for the database object.
+message DatabaseObjectSpec {
+  string protocol = 1;
+  string database_service_name = 2;
+  string object_kind = 3;
+  string database = 4;
+  string schema = 5;
+  string name = 6;
+}
+```
+
+All of the above fields are optional, and more may be added in the future if
+needed. The database-specific implementation is responsible for populating these
+fields from the database schema.
+
+### Database object permissions
+
+The permissions for particular objects are defined in a role using the new
+`db_permissions` field, found under `spec.allow` and `spec.deny` respectively.
+Each permission specifies:
+
+- A list of labels the database object must match.
+- A list of permissions that will be given to the object for the user.
+
+As an example, the role `db_read_non_confidential` allows access to tables and
+views with a label `confidential: false` and explicitly disallows any access to
+those labeled `confidential: true`.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: db_read_non_confidential
+spec:
+  allow:
+    db_permissions:
+      - match:
+          confidential: 'false'
+          kind:
+            - table
+            - view
+        permissions:
+          - SELECT
+          - INSERT
+          - UPDATE
+  deny:
+    db_permissions:
+      - match:
+          confidential: 'true'
+        permissions:
+          - '*'
```

Comment:
Which of these fields support wildcard and which don't? Is it only the `permissions` in `deny` supports wildcard? Thanks.

Reply:
Yep, only `deny` allows you to use wildcard, explicitly for the purpose of denying all possible permissions.